### PR TITLE
Expression indexes

### LINKED ISSUE #11423: sql: remove expression index session and cluster setting
PR: https://github.com/cockroachdb/cockroach/pull/68807

From release notes:
> Indexes on expressions can now be created. These indexes can be used to satisfy queries that contain filters with identical expressions. For example, `SELECT * FROM t WHERE a + b = 10` can utilize an index like `CREATE INDEX i ON t ((a + b))`. [#68807][#68807]

### PR DESCRIPTION
**Pull request by stbof**
Fixes #11423.

### CODE REVIEW

File: v21.2/expression-indexes.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+~~~
+
+
+## Known limitations
+
+Expression indexes have the following limitations:
+
+- The expression cannot reference columns outside the index's table.
+- The expression cannot reference [computed columns](computed-columns.html). [Tracking issue](https://github.com/cockroachdb/cockroach/issues/67900)
+- Functional expression output must be determined by the input arguments. For example, you can't use function `now()` to create an index because its output depends on more than just the function arguments.
```
Comment (mgartner) [inline v21.2/expression-indexes.md:129]: nit: "you can't use function" => "you can't use **the** function"

**Review by mgartner — APPROVED** [review]
LGTM

**Review by vy-ton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/11968)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[v21.2/computed-columns.md, line 11 at r6](https://reviewable.io/reviews/cockroachdb/docs/11968#-Mo-y_9h6Xn0d6PazIl9:-Mo-y_9h6Xn0d6PazIlA:b-bcw90b) ([raw file](https://github.com/cockroachdb/docs/blob/45654b874c864319457cf61f5e80e5d541124d1e/v21.2/computed-columns.md#L11)):*
> ```Markdown
> A _stored computed column_ (set with the `STORED` SQL keyword) is calculated when a row is inserted or updated, and stores the resulting value of the scalar expression in the primary index similar to a regular column.
> 
> A _virtual computed column_ (set with the `VIRTUAL` SQL keyword) is not stored, and the value of the scalar expression is computed during at query-time as needed.
> ```

Minor: remove `during`
___
*[v21.2/expression-indexes.md, line 9 at r6](https://reviewable.io/reviews/cockroachdb/docs/11968#-Mo-zR24AT9_ihGaKRV3:-Mo-zR24AT9_ihGaKRV4:bhxdr2c) ([raw file](https://github.com/cockroachdb/docs/blob/45654b874c864319457cf61f5e80e5d541124d1e/v21.2/expression-indexes.md#L9)):*
> ```Markdown
> <span class="version-tag">New in v21.2</span>
> 
> An _expression index_ is an index created by applying an [expression](scalar-expressions.html) to a column. For example, to facilitate fast, case insensitive lookups of user names you could create an index by applying the function `lower` to the `name` column: `CREATE INDEX users_name_idx ON users (lower(name))`. The value of the expression is stored only in the expression index, not in the primary family index.
> ```

Might want to mention to can reference multiple columns in an expression index
___
*[v21.2/expression-indexes.md, line 99 at r6](https://reviewable.io/reviews/cockroachdb/docs/11968#-Mo0-dJxDBA3yJ3XeNS2:-Mo0-dJxDBA3yJ3XeNS3:b-kud1c6) ([raw file](https://github.com/cockroachdb/docs/blob/45654b874c864319457cf61f5e80e5d541124d1e/v21.2/expression-indexes.md#L99)):*
> ```Markdown
> {% include_cached copy-clipboard.html %}
> ~~~sql
> > SELECT jsonb_pretty(user_profile) FROM users@timestamp_idx WHERE user_profile->'birthdate' = '"2011-11-07"';
> ```

Here you're using an index hint.
For a better recipe, you could display the `EXPLAIN` output for a query before and after the expression index creation, which would show the optimizer using the newly created index.
___
*[v21.2/hash-sharded-indexes.md, line 9 at r6](https://reviewable.io/reviews/cockroachdb/docs/11968#-Mo00QvtB4mnRwqqoRLo:-Mo00QvtB4mnRwqqoRLp:b-bwpp7q) ([raw file](https://github.com/cockroachdb/docs/blob/44b6215738320278d5701e23c735faa45d76e62b/v21.2/hash-sharded-indexes.md#L9)):*
> ```Markdown
> If you are working with a table that must be indexed on sequential keys, you should use **hash-sharded indexes**. Hash-sharded indexes distribute sequential traffic uniformly across ranges, eliminating single-range hotspots and improving write performance on sequentially-keyed indexes at a small cost to read performance.
> 
> {% include {{ page.version.version }}/misc/experimental-warning.md %}
> ```

This should not be removed, hash-sharded indexes are still experimental
___
*[v21.2/jsonb.md, line 78 at r6](https://reviewable.io/reviews/cockroachdb/docs/11968#-Mo01BnG2XM7-OacF5VC:-Mo01BnG2XM7-OacF5VD:b-b7l78o) ([raw file](https://github.com/cockroachdb/docs/blob/44b6215738320278d5701e23c735faa45d76e62b/v21.2/jsonb.md#L78)):*
> ```Markdown
> If the execution of a [join](joins.html) query exceeds the limit set for [memory-buffering operations](vectorized-execution.html#disk-spilling-operations) (i.e., the value set for the `sql.distsql.temp_storage.workmem` [cluster setting](cluster-settings.html)), CockroachDB will spill the intermediate results of computation to disk. If the join operation spills to disk, and at least one of the columns is of type `JSON`, CockroachDB returns the error `unable to encode table key: *tree.DJSON`. If the memory limit is not reached, then the query will be processed without error.
> 
> For details, see [tracking issue](https://github.com/cockroachdb/cockroach/issues/35706).
> ```

Why did we remove the tracking issue?

File: v21.2/computed-columns.md [hunk 2]
```diff
@@ -4,7 +4,11 @@ summary: A computed column exposes data generated by an expression included in t
 toc: true
 ---
 
-A computed column exposes data generated from other columns by a [scalar expression](scalar-expressions.html) included in the column definition. A stored computed column (set with the `STORED` SQL keyword) is calculated when a row is inserted or updated, and stores the resulting value of the scalar expression in the primary index similar to a regular column. A virtual computed column (set with the `VIRTUAL` SQL keyword) is not stored, and the value of the scalar expression is computed during queries as needed.
+A _computed column_ exposes data generated from other columns by a [scalar expression](scalar-expressions.html) included in the column definition.
+
+A _stored computed column_ (set with the `STORED` SQL keyword) is calculated when a row is inserted or updated, and stores the resulting value of the scalar expression in the primary index similar to a regular column.
+
+A _virtual computed column_ (set with the `VIRTUAL` SQL keyword) is not stored, and the value of the scalar expression is computed during at query-time as needed.
```
Comment (ianjevans) [inline v21.2/computed-columns.md:11]: "is computed at query-time as needed"

File: v21.2/expression-indexes.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+An inverted, partial, multi-column index, where one column is defined with an expression:
+{% include_cached copy-clipboard.html %}
+~~~sql
+CREATE INVERTED INDEX ON t (lower(s), i, j) WHERE b;
+~~~
+
+### Use an expression to index a field in a `JSONB` column
+
+You can use an [expression](expression-indexes.html) in an index definition to index a field in a JSON column. You can also use an expression to create an [inverted index](inverted-indexes.html) on a subset of the JSON column.
```
Comment (ianjevans) [inline v21.2/expression-indexes.md:75]: The "expression" link is self-referential.

File: v21.2/expression-indexes.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+The following example creates a table of users with a JSON object in `user_profile` column and then an index on the `birthdate` field in that column:
```
Comment (ianjevans) [inline v21.2/expression-indexes.md:77]: "JSON object in the `user_profile` column."

File: v21.2/jsonb.md [hunk 5]
```diff
@@ -47,7 +43,7 @@ Examples:
 
 ## Size
 
-The size of a `JSONB` value is variable, but it's recommended to keep values under 1 MB to ensure performance. Above that threshold, [write amplification](https://en.wikipedia.org/wiki/Write_amplification) and other considerations may cause significant performance degradation.
+The size of a `JSONB` value is variable, but we recommend that you keep values under 1 MB to ensure satisfactory performance. Above that threshold, [write amplification](https://en.wikipedia.org/wiki/Write_amplification) and other considerations may cause significant performance degradation.
```
Comment (ianjevans) [inline v21.2/jsonb.md:46]: Remove the link to the Wikipedia entry on write amplification, as that article is about SSD storage. Refer to [this Slack conversation](https://cockroachlabs.slack.com/archives/C01PHNMUFLN/p1636058091052900) for more discussion on defining read/write amplification.

File: v21.2/scalar-expressions.md [hunk 6]
```diff
@@ -27,30 +26,28 @@ They are described further in section [SQL Constants](sql-constants.html).
...
-  *This is a CockroachDB SQL extension.*
+    {{site.data.alerts.callout_info}}This is a CockroachDB SQL extension.{{site.data.alerts.end}}
 
-  {{site.data.alerts.callout_danger}}
-  Ordinal references should be used with care in production
-  code! During schema updates, column ordinal positions can change and
-  invalidate existing queries that use ordinal positions based on a
-  previous version of the schema.
-  {{site.data.alerts.end}}
+    {{site.data.alerts.callout_danger}}
```
Comment (ianjevans) [inline v21.2/scalar-expressions.md:45]: nit: the line breaks in this section are weird. Actually, the line breaks within the page could all be cleaned up.

**Review by ianjevans — COMMENTED** [review]
Just some small fixes that could be addressed in a follow-up PR.

**Comment by stbof** [comment]
I am working on a follow up PR today. Thanks for the reviews all!

**Pull request merged**